### PR TITLE
SNOW-631792 adding EXECUTE AS and tests for correctness

### PR DESCRIPTION
**Pull request by sfc-gh-mkeller**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-631792

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   In this PR I add `EXECUTE AS` clause to stored procedures.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 1]
```diff
@@ -659,14 +661,15 @@ def create_python_udf_or_sp(
     )
 
     create_query = f"""
-CREATE {"OR REPLACE " if replace else ""}
+CREATE{" OR REPLACE " if replace else ""}
```
Comment (sfc-gh-mkeller) [inline src/snowflake/snowpark/_internal/udf_utils.py:664]: This is an unrelated change, it just removes a space from the generated SQL command

File: src/snowflake/snowpark/stored_procedure.py [hunk 2]
```diff
@@ -390,6 +397,7 @@ def register(
             replace,
             parallel,
             statement_params=statement_params,
+            execute_as=execute_as,
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/stored_procedure.py:400]: Should we check the value of `execute_as` in `_do_register_sp` to be either `owner` or `caller`? This way it also makes negative test easier.

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 3]
```diff
@@ -659,14 +667,14 @@ def create_python_udf_or_sp(
...
 {"TEMPORARY" if is_temporary else ""} {object_type.value} {object_name}({sql_func_args})
 {return_sql}
 LANGUAGE PYTHON
 RUNTIME_VERSION=3.8
 {imports_in_sql}
 {packages_in_sql}
-HANDLER='{handler}'
+HANDLER='{handler}'{execute_as_sql}
```
[the remark this answers is not in the document]
Comment (sfc-gh-mkeller) [reply src/snowflake/snowpark/_internal/udf_utils.py:677]: But this function is also used for UDTFs, which do not have `EXECUTE AS`.
So if we do that UDTFs will have an empty line in their SQL commands. Is that okay?

File: src/snowflake/snowpark/functions.py [hunk 4]
```diff
@@ -3399,6 +3401,7 @@ def sproc(
     session: Optional["snowflake.snowpark.Session"] = None,
     parallel: int = 4,
     statement_params: Optional[Dict[str, str]] = None,
+    execute_as: te.Literal["caller", "owner"] = "owner",
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/functions.py:3404]: hmm what's the different between Literal from typing and typing_extensions?

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/functions.py [hunk 5]
```diff
@@ -3454,6 +3457,8 @@ def sproc(
             command. The default value is 4 and supported values are from 1 to 99.
             Increasing the number of threads can improve performance when uploading
             large stored procedure files.
+        execute_as: What permissions should the procedure have while executing. This
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/functions.py:3460]: add a link https://docs.snowflake.com/en/sql-reference/stored-procedures-rights.html?

File: src/snowflake/snowpark/functions.py (hunk 4 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/snowpark/functions.py:3404]: Nothing, I thought it was introduced after 3.8. [I was wrong tho](https://docs.python.org/3/library/typing.html#typing.Literal).
> New in version 3.8.

I'll update my code.

File: src/snowflake/snowpark/_internal/udf_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/udf_utils.py:677]: Ah thanks for letting me know. I'm not aware of this. Maybe let's add a comment where we define `execute_as_sql`? There might be other people getting confused by the same :)

I guess today we might face the same for `imports_in_sql` and `packages_in_sql`, we probably could refactor those later as well.

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/functions.py [hunk 6]
```diff
@@ -3454,6 +3456,9 @@ def sproc(
...
+            supports caller, or owner for now. See https://docs.snowflake.com/en/sql-reference/stored-procedures-rights.html
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/functions.py:3460]:
```suggestion
            supports caller, or owner for now. See `owner and caller rights <https://docs.snowflake.com/en/sql-reference/stored-procedures-rights.html>`_
```
and let's have this link for all docstrings having this arg?

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 7]
```diff
@@ -21,6 +21,7 @@
 )
 
 import cloudpickle
+from typing_extensions import Literal
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/udf_utils.py:24]:
```suggestion
from typing import Literal
```

File: src/snowflake/snowpark/stored_procedure.py [hunk 8]
```diff
@@ -7,6 +7,8 @@
 from types import ModuleType
 from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, Union
 
+from typing_extensions import Literal
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/stored_procedure.py:10]:
```suggestion
from typing import Literal
```

File: src/snowflake/snowpark/functions.py [hunk 9]
```diff
@@ -169,6 +169,7 @@
     <BLANKLINE>
 """
 import functools
+import typing
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/functions.py:172]: just for consistency, we can do `from typing import Literal` as well ?

File: src/snowflake/snowpark/_internal/udf_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mkeller) [reply src/snowflake/snowpark/_internal/udf_utils.py:677]: I'll add a comment, no problem!

File: src/snowflake/snowpark/functions.py (hunk 9 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/snowpark/functions.py:172]: Now, that we cannot. `flake8` does some special checking for type-hints to be pointing at valid things. However; if we did `from typing import Literal` then we bring in `Literal` into a local namespace and it's impossible for tooling to know that it was imported from `typing` originally.

You can see the my bug report at https://github.com/PyCQA/pyflakes/issues/725 and once I dug in deeper into the duplicated issue I must say that I agree with this
